src/hotspot/share/compiler/compilerOracle.cpp

```diff
@@ -410,15 +415,16 @@ bool CompilerOracle::should_break_at(const methodHandle& method) {
 }
 
 bool CompilerOracle::should_blackhole(const methodHandle& method) {
-  if (check_predicate(CompileCommand::Blackhole, method)) {
-    if (method->result_type() == T_VOID) {
-      return true;
-    } else {
-      warning("blackhole compile command only works for methods with void type: %s",
-              method->name_and_sig_as_C_string());
-    }
+  if (!check_predicate(CompileCommand::Blackhole, method)) {
+    return false;
   }
-  return false;
+  if (method->result_type() != T_VOID) {
+    warning("Blackhole compile option only works for methods with void type: %s",
+            method->name_and_sig_as_C_string());
+    return false;
+  }
+  guarantee(UnlockDiagnosticVMOptions, "Checked during initial parsing");
```

Comment:
You can move it up right after `check_predicate`: `check_predicate(CompileCommand::Blackhole, method)` is allowed to return `true` iff `UnlockDiagnosticVMOptions` is set. 

Not sure `guarantee` is well-justified here, IMO `assert` would suffice.

Reply:
`UnlockDiagnosticVMOptions` is `trueInDebug`, so `assert` is useless. We really care about "release" mode here.

Reply:
Moved the `guarantee` closer to `check_predicate`.